File: src/main/java/org/cmjava2023/BytecodeFromClassfileModel.java
package org.cmjava2023;

import java.lang.reflect.Array;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HexFormat;
import java.util.List;

public class BytecodeFromClassfileModel {

    static final int magicNumber = 0xCAFEBABE;
    static final short minorVersion = 0x0;
    static final short majorVersion = 0x00000034;

    public Byte[] Generate(ClassfileModel model){


        var array = new ArrayList<Byte>(toObjects(ByteBuffer.allocate(4).putInt(magicNumber).array()));
        array.addAll(toObjects(ByteBuffer.allocate(2).putShort(minorVersion).array()));
        array.addAll(toObjects(ByteBuffer.allocate(2).putShort(majorVersion).array()));

        // TODO: Fix java
        var result = new Byte[array.size()];
        array.toArray(result);
        return result;
    }

    List<Byte> toObjects(byte[] bytesPrim) {

        var bytes = new ArrayList<Byte>();
        for(var item : bytesPrim){
            bytes.add(item);
        }
        return bytes;
    }
}

File: src/main/kotlin/org/cmjava2023/BytecodeFromClassfileModel.kt
package org.cmjava2023

import org.cmjava2023.util.ByteListUtil.Companion.add
import org.cmjava2023.classfilespecification.ClassfileModel
import org.cmjava2023.util.AccessModifierUtil.Companion.bitwiseOrCombine


class BytecodeFromClassfileModel {

    fun generate(model: ClassfileModel): ByteArray {
        val classFileBytes = ConstantPoolToByteList().mapToClassFileBytes(model.constantPool, model.methodDefinitions)

        val result: MutableList<Byte> = mutableListOf()
        result.add(MAGIC_NUMBER)
        result.add(MINOR_VERSION)
        result.add(MAJOR_VERSION)
        result.add(classFileBytes.constantPoolItemCount)
        result.addAll(classFileBytes.constantPoolBytes)
        result.add(model.classClassAccessModifiers.map { it.value }.bitwiseOrCombine())
        result.add(INDEX_OF_THIS_CLASS_IN_CONSTANT_POOL)
        result.add(INDEX_OF_SUPER_CLASS_IN_CONSTANT_POOL)
        result.add(model.numberOfInterfaces)
        result.addAll(model.interfaceDefinitions)
        result.add(model.numberOfFields)
        result.addAll(model.fieldDefinitions)
        result.add(classFileBytes.methodInfoCount)
        result.addAll(classFileBytes.methodInfosBytes)
        result.add(model.attributesCount)
        result.addAll(model.attributeDefinitions)

        return result.toByteArray()
    }

    companion object {
        const val MAGIC_NUMBER: UInt = 0xCAFEBABEu
        const val MINOR_VERSION: UShort = 0u
        const val MAJOR_VERSION: UShort = 50u

        const val INDEX_OF_THIS_CLASS_IN_CONSTANT_POOL: UShort = 2u
        const val INDEX_OF_SUPER_CLASS_IN_CONSTANT_POOL: UShort = 4u
    }
}

